pkg/server/plugin/upstreamauthority/gcpca/gcpca.go

```diff
@@ -0,0 +1,332 @@
+	}
+
+	// Swap out the current configuration with the new configuration
+	p.setConfig(config)
+
+	return &plugin.ConfigureResponse{}, nil
+}
+
+func (p *Plugin) GetPluginInfo(ctx context.Context, req *plugin.GetPluginInfoRequest) (*plugin.GetPluginInfoResponse, error) {
+	return &plugin.GetPluginInfoResponse{}, nil
+}
+
+func (p *Plugin) getConfig() (*Config, error) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	if p.c == nil {
+		return nil, pluginErr.New("not configured")
+	}
+
+	return p.c, nil
+}
+
+func (p *Plugin) setConfig(c *Config) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.c = c
+}
+
+func getClient(ctx context.Context) (*pcaapi.CertificateAuthorityClient, error) {
+	creds, err := google.FindDefaultCredentials(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	pcaClient, err := pcaapi.NewCertificateAuthorityClient(ctx, option.WithCredentials(creds))
+	if err != nil {
+		return nil, err
+	}
+
+	return pcaClient, nil
+}
+
+func pemToX509(pemStr string) (*x509.Certificate, error) {
+	block, rest := pem.Decode([]byte(pemStr))
+	if len(rest) > 0 {
+		return nil, errors.New("Unexpected trailer in certificate of size " + fmt.Sprint(len(rest)))
+	}
+
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return cert, nil
+}
+
+func loadCertificateAuthorities(ctx context.Context, client *pcaapi.CertificateAuthorityClient, spec CertificateAuthoritySpec) ([]*privatecapb.CertificateAuthority, error) {
+	allCerts := make([]*privatecapb.CertificateAuthority, 0)
+	certIt := client.ListCertificateAuthorities(ctx, &privatecapb.ListCertificateAuthoritiesRequest{
+		Parent: spec.caParentPath(),
+		Filter: "labels." + spec.LabelKey + ":" + spec.LabelValue,
+	})
+
+	p := iterator.NewPager(certIt, 20, "")
+	for {
+		var page []*privatecapb.CertificateAuthority
+
+		nextPageToken, err := p.NextPage(&page)
+		if err != nil {
+			return nil, err
+		}
+
+		allCerts = append(allCerts, page...)
+		if nextPageToken == "" {
+			break
+		}
+	}
+
+	return allCerts, nil
+}
+
+func (p *Plugin) mintX509CA(ctx context.Context, csr []byte, preferredTtl int32) (*upstreamauthority.MintX509CAResponse, error) {
+	csrParsed, err := x509.ParseCertificateRequest(csr)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse CSR: %v", err)
+	}
+
+	validity := time.Second * time.Duration(preferredTtl)
+
+	pcaClient, err := getClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	allCertRoots, err := loadCertificateAuthorities(ctx, pcaClient, p.c.RootSpec)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to load root CAs: %v.", err)
+	}
+	if len(allCertRoots) == 0 {
+		return nil, fmt.Errorf("No certificate authorities found with label pair %s:%s.", p.c.RootSpec.LabelKey, p.c.RootSpec.LabelValue)
+	}
+
+	// The certs get returned in issuance order, so just grab the last one and use that.
+	chosenCA := allCertRoots[len(allCertRoots)-1]
+	trustBundle := allCertRoots[:len(allCertRoots)-1]
+	if p.c.TrustBundleSpec.LabelKey != "" {
```

Comment:
The TrustBundleSpec actually just allows you to tell your SPIRE-managed workloads about other certs they should be able to reason about as roots.  The resultant trust bundle would allow for both the cert rotation and the 'other cert' trust.  Does that make sense?

Reply:
Thanks, that makes sense. Are we planning to recommend using the `TrustBundleSpec` for rotation, or should users just use the rotation flow, using labeled CAs, that we've described above?

Reply:
I think just the rotation flow and labels.